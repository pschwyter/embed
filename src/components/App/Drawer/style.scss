@import "~style/utilities/colors.scss";

.ADA-CHAPERONE-Drawer {
	&__iframe {
		display: block;
		height: 100vh;
		width: 100%;
		max-width: 375px;
		position: fixed;
		bottom: 0;
		right: 0;
		box-sizing: border-box;
		box-shadow: -1px 0 0 rgba(0, 0, 0, 0.1), -3px 0 9px rgba(0, 0, 0, 0.15);
		pointer-events: all;
		transform: translate(0, 0);
		-webkit-transition: -webkit-transform 200ms ease, opacity 200ms ease;
		transition: transform 200ms ease, opacity 200ms ease;
		overflow: initial;
		border: none;
		outline: none;
		width: 100%;
		height: 100%;
		box-sizing: border-box;
		z-index: 9999;
		background-color: $color-white;
	}

	&--hidden &__iframe {
		transform: translate(375px, 0);
		-webkit-transform: translate(375px, 0);
		-moz-transform: translate(375px, 0);
		opacity: 0;
	}

	&__mask {
		position: fixed;
		width: 100%;
		height: 100%;
		top: 0;
		left: 0;
		background-color: rgba(black, 0.16);
		z-index: 9998;
		transition: all 200ms ease;		
		-webkit-backdrop-filter: grayscale(100%);
		display: block;
		visibility: visible;
		opacity: 1;
		pointer-events: all;
	}
	
	&--hidden &__mask {
		display: none;
		opacity: 0;
		visibility: hidden;
		pointer-events: none;
		-webkit-backdrop-filter: grayscale(0);
	}

	// &--hidden#{&}--mobile-overlay {
	// 	transform: translate(100vw, 0);
	// 	-webkit-transform: translate(100vw, 0);
	// 	-moz-transform: translate(100vw, 0);
	// }

	// &--mobile-overlay {
	// 	max-width: none;
	// 	height: 100%;
	// 	min-height: 100%;
	// }
}
