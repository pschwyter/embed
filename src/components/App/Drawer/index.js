import { Component } from "preact";
import classnames from "classnames";
import { isIE9OrBelow } from "services/browsers";
import "./style.scss";

export default class Drawer extends Component {
  /**
   * @param {Object} props
   */
  constructor(props) {
    super(props);

    this.state = {
      drawerHasBeenOpened: false
    };

    this.isIE9OrBelow = isIE9OrBelow();
  }

  /**
   * @param {Object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    const { isDrawerOpen } = this.props;
    const { drawerHasBeenOpened } = this.state;

    if (isDrawerOpen !== nextProps.isDrawerOpen && !drawerHasBeenOpened) {
      this.setState({
        drawerHasBeenOpened: true
      });
    }
  }

  /**
   * @param {Object} props
   * @returns {ReactElement}
   */
  render(props) {
    const {
      handle,
      isDrawerOpen,
      openChat,
      chatURL,
      useMobileOverlay
    } = props;
    const { drawerHasBeenOpened } = this.state;

    return (
      <div
        className={classnames(
          "ada-chaperone-drawer",
          {
            "ada-chaperone-drawer--hidden": !isDrawerOpen,
            "ada-chaperone-drawer--isIE9": this.isIE9OrBelow,
            "ada-chaperone-drawer--mobile-overlay": useMobileOverlay
          }
        )}
      >
        <div
          className="ada-chaperone-drawer__mask"
          onClick={openChat}
        />
        <div className="ada-chaperone-drawer__iframe-container">
          {drawerHasBeenOpened && (
            <iframe
              className="ada-chaperone-drawer__iframe-container__iframe"
              src={chatURL}
              title={`${handle} chat support`}
            />
          )}
        </div>
      </div>
    );
  }
}
