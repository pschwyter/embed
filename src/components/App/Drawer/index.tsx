import classnames from "classnames";
import Client from "models/Client";
import { Component, h } from "preact";
import { isIE9OrBelow } from "services/browsers";
import IFrame from "../IFrame";
import "./style.scss";

interface InterfaceDrawer {
  client: Client,
  handle: string,
  isDrawerOpen: boolean,
  iframeRef?: HTMLIFrameElement,
  chatURL: string,
  useMobileOverlay: boolean,
  drawerHasBeenOpened: boolean,
  toggleChat(): void,
  setIFrameRef(ref: HTMLIFrameElement): void,
  setIFrameLoaded(): void
}

export default class Drawer extends Component<InterfaceDrawer> {
  isIE9OrBelow: boolean;

  constructor(props: InterfaceDrawer) {
    super(props);

    this.isIE9OrBelow = isIE9OrBelow();
  }

  render() {
    const {
      isDrawerOpen,
      toggleChat,
      useMobileOverlay,
      drawerHasBeenOpened
    } = this.props;

    return (
      <div
        className={classnames(
          "ada-chaperone-drawer",
          {
            "ada-chaperone-drawer--hidden": !isDrawerOpen,
            "ada-chaperone-drawer--isIE9": this.isIE9OrBelow,
            "ada-chaperone-drawer--mobile-overlay": useMobileOverlay
          }
        )}
      >
        <div
          className="ada-chaperone-drawer__mask"
          onClick={toggleChat}
          role="button"
        />
        <div
          className="ada-chaperone-drawer__iframe-container"
          role="dialog"
          aria-modal={isDrawerOpen}
          aria-hidden={!isDrawerOpen}
        >
          {drawerHasBeenOpened && (
            <IFrame
              {...this.props}
            />
          )}
        </div>
      </div>
    );
  }
}
