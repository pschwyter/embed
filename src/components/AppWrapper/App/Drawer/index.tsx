import classnames from "classnames";
import { Component, h } from "preact";
import { isIE9OrBelow } from "services/browsers";
import IFrame from "../IFrame";
import "./style.scss";

interface InterfaceDrawer {
  handle: string,
  isDrawerOpen: boolean,
  iframeRef?: HTMLIFrameElement,
  chatURL: string,
  useMobileOverlay?: boolean,
  hideMask?: boolean,
  drawerHasBeenOpened: boolean,
  introShown: boolean,
  toggleChat(): void,
  setIFrameRef(ref: HTMLIFrameElement): void,
  setIFrameLoaded(): void,
  transitionEndCallback(): void
}

export default class Drawer extends Component<InterfaceDrawer> {
  isIE9OrBelow: boolean;
  drawerRef: any;

  constructor(props: InterfaceDrawer) {
    super(props);

    this.isIE9OrBelow = isIE9OrBelow();
  }

  componentDidMount() {
    this.drawerRef.addEventListener("transitionend", this.props.transitionEndCallback, false);
  }

  componentWillUnmount() {
    this.drawerRef.removeEventListener("transitionend", this.props.transitionEndCallback, false);
  }

  render() {
    const {
      hideMask,
      toggleChat,
      isDrawerOpen,
      useMobileOverlay,
      drawerHasBeenOpened
    } = this.props;

    return (
      <div
        className={classnames(
          "ada-embed-drawer",
          {
            "ada-embed-drawer--hidden": !isDrawerOpen,
            "ada-embed-drawer--isIE9": this.isIE9OrBelow,
            "ada-embed-drawer--mobile-overlay": useMobileOverlay
          }
        )}
        ref={c => this.drawerRef = c}
      >
        {!hideMask &&
          <div
            className="ada-embed-drawer__mask"
            onClick={toggleChat}
            role="button"
          />
        }
        <div
          className="ada-embed-drawer__iframe-container"
          role="dialog"
          aria-modal={isDrawerOpen}
          aria-hidden={!isDrawerOpen}
        >
          {drawerHasBeenOpened && (
            <IFrame
              {...this.props}
            />
          )}
        </div>
      </div>
    );
  }
}
