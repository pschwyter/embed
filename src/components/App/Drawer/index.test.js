/* eslint-env node, jest */

import "jsdom-global/register";
import { shallow } from "preact-render-spy";
import Drawer from "./index";

function setup(propsOverride = {}) {
  const defaultsProps = {
    client: {
      tint: "#000"
    },
    handle: "nic",
    isDrawerOpen: true,
    chatURL: "https://nic.ca.ada.support/chat/",
    useMobileOverlay: true,
    drawerHasBeenOpened: true,
    openChat: jest.fn(),
    setIFrameRef: jest.fn()
  };

  const props = Object.assign(defaultsProps, propsOverride);
  const PRSWrapper = shallow(<Drawer {...props} />);

  return {
    props,
    PRSWrapper
  };
}

describe("<Drawer />", () => {
  it("should match snapshot", () => {
    const { PRSWrapper } = setup();

    expect(PRSWrapper).toMatchSnapshot();
  });

  it("should hide the drawer when isDrawerOpen is set to false", () => {
    const { PRSWrapper } = setup({
      isDrawerOpen: false
    });

    expect(PRSWrapper.find(".ada-chaperone-drawer--hidden").length).toBe(1);
  });

  it("should use special styling when useMobileOverlay is true", () => {
    const { PRSWrapper } = setup({
      useMobileOverlay: true
    });

    expect(PRSWrapper.find(".ada-chaperone-drawer--mobile-overlay").length).toBe(1);
  });
});
