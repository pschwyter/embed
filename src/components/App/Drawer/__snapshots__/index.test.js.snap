// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should match snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<div class="ada-chaperone-drawer ada-chaperone-drawer--mobile-overlay">
  <div
    class="ada-chaperone-drawer__mask"
    onClick={[Function mockConstructor]}
    role="button"
  >
  </div>
  <div
    class="ada-chaperone-drawer__iframe-container"
    role="dialog"
    aria-modal={true}
    aria-hidden={false}
  >
    <IFrame
      client={
        Object {
          "tint": "#000"
        }
      }
      handle="nic"
      isDrawerOpen={true}
      chatURL="https://nic.ca.ada.support/chat/"
      useMobileOverlay={true}
      drawerHasBeenOpened={true}
      toggleChat={[Function mockConstructor]}
      setIFrameRef={[Function mockConstructor]}
    >
    </IFrame>
  </div>
</div>

`;
